File: Set Card Game/JavaScript/src/game.js
import Deck from './deck.js'
import Card from './card.js'
import Stats from './stats.js'
import { Player, ComputerPlayer } from './player.js'

// Game will hold all the logic for manipulating data objects (e.g. deck, cards on board, players)
export default class Game {
    constructor() {
        this.deck = new Deck();
        this.array_of_player = [];
        this.active_cards = [];
    }

    // if a user cannot find a set, reshuffle cards
    reshuffle(stats_map) {
        while (this.active_cards.length > 0) {
            this.deck.new_cards.push(this.active_cards.pop());
        }
        // track how many times the user reshuffles
        Stats.count_num_reshuffles(stats_map);
        this.deck.shuffle();
        this.deal(stats_map);
        // we don't wat to increment count of deals when reshuffling
        Stats.decrement_number_deals(stats_map);
    }

    // name is a string
    add_player(name) {
        this.array_of_player.push(new Player(name));
    }

    // name is a string difficulty is passed by user
    add_cpu_player(name, difficulty) {
        this.array_of_player.push(new ComputerPlayer(name, difficulty));
    }

    // adds the next card in the game's deck to the active cards
    add_card() {
        this.active_cards.push(this.deck.next_card());
    }

    // deal cards to board
    deal(stats_map) {
        while (this.active_cards.length < 12 && this.deck.new_cards.length !== 0) {
            this.add_card();
        }
        // count total number of deals
        Stats.count_num_deals(stats_map);
    }

    // removes the cards that made a set
    delete_cards(card_index_1, card_index_2, card_index_3) {
        this.active_cards.splice(this.active_cards.findIndex(card => card.index === card_index_1), 1);
        this.active_cards.splice(this.active_cards.findIndex(card => card.index === card_index_2), 1);
        this.active_cards.splice(this.active_cards.findIndex(card => card.index === card_index_3), 1);
    }

    // finds a set in the active array
    find_set() {
        let i = 0;
        let j = i + 1;
        let k = j + 1;
        let setFound = false;
        let returnSet = [];
        while(i < this.active_cards.length  && !setFound) {
            while(j < this.active_cards.length && !setFound) {
                while(k < this.active_cards.length && !setFound) {
                    if(Card.is_set(this.active_cards[i], this.active_cards[j], this.active_cards[k])) {
                        setFound = true;
                        returnSet = [i,j,k];
                    }
                    k++;
                }
                j++;
                k = j + 1;
            }
            i++;
            j = i + 1;
        }
        return returnSet;
    }

    // checks if theres any sets left in the entire game
    any_set() {
        let all_cards = this.active_cards.concat(this.deck.new_cards);
        let i = 0;
        let j = i + 1;
        let k = j + 1;
        let setFound = false;
        while(i < all_cards.length  && !setFound) {
            while(j < all_cards.length && !setFound) {
                while(k < all_cards.length && !setFound) {
                    if(Card.is_set(all_cards[i], all_cards[j], all_cards[k])) {
                        setFound = true;
                    }
                    k++;
                }
                j++;
                k = j + 1;
            }
            i++;
            j = i + 1;
        }
        return setFound;
    }

    // runs the players guess
    set_checker(index1, index2, index3, player, stats_map) {
        if(Card.is_set(this.active_cards[index1], this.active_cards[index2], this.active_cards[index3])) {
            Stats.count_correct_answers(stats_map);
            // unsure how output will be
            player.consecutive_sets++;
            player.points += player.consecutive_sets;
            // Blake I think you have something to add for your stat map in this
            // Stats.count_correct_answers(stat_map);
            this.delete_cards(this.active_cards[index1].index, this.active_cards[index2].index, this.active_cards[index3].index);
            this.deal(stats_map);
            return true;
            // Stats.count_num_deals(stat_map);
        } else {
            // output again
            Stats.count_incorrect_answers(stats_map);
            player.consecutive_sets = 0;
            player.points--;
            return false;
        }
    }
}

File: Set Card Game/Ruby/game.rb
# frozen_string_literal: true

require_relative  'card'
require_relative  'deck'
require_relative  'player'
require_relative 'stats'
require 'colorize'
# Game class will maintain the deck and cards in play as well as all game logic
class Game
  attr_accessor :array_of_cards
  attr_accessor :deck
  attr_accessor :array_of_player
  attr_accessor :playable
  # Param array_of_player: an array that contains players' score.
  # Param array_of_cards: an array that contains active cards.
  # initialize the player array and the active cards array in the game
  def initialize
    @deck = Deck.new
    @array_of_player = []
    @array_of_cards = []
    @playable = true
  end

  # Param name: name of the player
  # add player into the game
  def add_player(name)
    @array_of_player.push(Player.new(name))
  end

  # Param name: name of the CPU player
  # Param difficulty: difficulty of the CPU player
  # add player into the game
  def add_cpu_player(name, difficulty)
    @array_of_player.push(ComputerPlayer.new(name, difficulty))
  end

  # deal appropriate number of cards to the board
  def deal
    until check_active_cards || @deck.new_cards.none?
      add_card(@deck.new_cards[0])
      @deck.next_card
    end
  end
  
  # check the number of active cards is greater than 12
  def check_active_cards
    @array_of_cards.length >= 12
  end

  # deals at most three additional cards as long as the deck is not empty
  def three_more
    i = 0
    until i == 3 || @deck.new_cards.none?
      add_card(@deck.new_cards[0])
      @deck.next_card
      i += 1
    end
  end

  # Finds a set in the array_of_cards and returns the indices of those cards or an empty array
  def find_set
    (0...(@array_of_cards.length - 1)).each do |i|
      (i + 1...(@array_of_cards.length - 1)).each do |j|
        (j + 1...(@array_of_cards.length - 1)).each do |k|
          return [i, j, k] if Card.is_set(@array_of_cards[i], @array_of_cards[j], @array_of_cards[k])
        end
      end
    end
    []
  end

  # check if there is a possible set in the remaining cards
  def check_set_all
    all_cards = @array_of_cards + @deck.new_cards
    (0...(all_cards.length - 1)).each do |i|
      (i + 1...(all_cards.length - 1)).each do |j|
        (j + 1...(all_cards.length - 1)).each do |k|
          return true if Card.is_set(all_cards[i], all_cards[j], all_cards[k])
        end
      end
    end
    false
  end

  # Param card: the card going to be removed
  # remove card from active cards array
  def remove_card(card)
    @array_of_cards.delete(card)
  end

  # Param card: the card going to be added
  # add card to active cards array
  def add_card(card)
    @array_of_cards.push(card)
  end

  # Param hint: number used to specify which one of the three hints you wish to push
  # Param hints: an array stores all three hints
  # return a hint for player if hint exists
  def hint_pusher(hint, hints)
      if hint == 0
        puts("\nA set can be made with card " + hints[hint].to_s + "\n\n")
      elsif hint == 1
        puts("\nA set can be made with card "  + hints[hint - 1].to_s + ' and card ' + hints[hint].to_s + "\n\n")
      elsif hint == 2
        puts("\nA set can be made with card "  + hints[hint - 2].to_s + ', card ' + hints[hint - 1].to_s + ', and card ' + hints[hint].to_s + "\n\n")
      else
        puts("\nA set can be made with card "  + hints[0].to_s + ', card ' + hints[1].to_s + ', and card ' + hints[2].to_s + "\n\n")
      end
    hint + 1
  end

=begin
  function will add three more cards to deck and one point to player if the set is correct, otherwise
  minus one point to the player.
=end
  def set_checker(card1, card2, card3, player, input)
    # if the given set is correct, remove set from the deck, add three new card and player point add 1
    if Card.is_set(card1, card2, card3)
      puts("\nCongrats, " + input + " that's a point!")
      player.consecutive_sets += 1
      player.points += player.consecutive_sets
      Stats.count_correct_answers
      remove_card(card1)
      remove_card(card2)
      remove_card(card3)
      deal
      Stats.count_num_deals
      while still_playable && find_set.length != 3
        three_more
        Stats.count_num_extra_deals
      end
    # if the given set is a wrong set, then player point minus one.
    else
      puts("\nSorry, " + input + ' that is incorrect. You lose one point.')
      player.consecutive_sets = 0
      player.points -= 1
      Stats.count_incorrect_answers
    end
    print_scores
    sleep(5)
    system 'clear'
    display_cards(@array_of_cards)
  end
  
  # check if the game is still playable
  def still_playable
    @playable = (@deck.new_cards.size + @array_of_cards.size).positive? && check_set_all
  end

  # Param sec, time you hope to count down
  # count down timer
  def timer_count(sec)
    (1..sec).each do |i|
      sleep 1
      i
    end
  end
    
  # prinr the score of the player
  def print_scores
    puts("\n")
    for player in @array_of_player
      puts(player.name + ": " + player.points.to_s)
    end
  end
end

